pkg/services/ngalert/schedule/fetcher.go

```diff
@@ -3,15 +3,110 @@ package schedule
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 )
 
+// RulesStore is a store that provides alert rules for scheduling
+type RulesStore interface {
+	GetAlertRulesKeysForScheduling(ctx context.Context) ([]models.AlertRuleKeyWithVersion, error)
+	GetAlertRulesForScheduling(ctx context.Context, query *models.GetAlertRulesForSchedulingQuery) error
+}
+
+type FetcherCfg struct {
+	ReloadInterval       time.Duration
+	DisableGrafanaFolder bool
+}
+
+// BackgroundFetcher periodically refreshes rules and folders in the background.
+type BackgroundFetcher struct {
+	cfg       FetcherCfg
+	ticker    *time.Ticker
+	ruleStore RulesStore
+	metrics   *metrics.Scheduler
+	logger    log.Logger
+
+	mu                 sync.Mutex
+	latestRules        []*models.AlertRule
+	latestFolderTitles map[models.FolderKey]string
```

Comment:
I think there's a case for merging this with the scheduler's `schedulableAlertRules` registry in the future. I didn't tackle that for now, to keep the PR scoped.